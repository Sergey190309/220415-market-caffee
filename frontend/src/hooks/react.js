import {useState, useEffect, useRef, useCallback } from 'react'

export const useAppState = useState
export const useAppEffect = useEffect
export const useAppRef = useRef
export const useAppCallback = useCallback
