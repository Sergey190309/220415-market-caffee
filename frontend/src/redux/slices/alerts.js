import { createSlice } from '@reduxjs/toolkit'

export const initialState = {
  /**
   * state structure:
   * alerts: [
   *   {
   *     message: 'mockMessage',
   *     alertType: 'info',
   *     timeout: 3000,
   *     id: 'mockId'
   *   }
   * ]
   */
  alerts: []
}

const alertsSlice = createSlice({
  name: 'alerts',
  initialState,
  reducers: {
    startAlert: (state, { payload }) => {
      state.alerts.push(payload)
    },
    removeAlert: (state, { payload }) => {
      // console.log('alertsSlicer, removeAlert, payload ->', payload);
      state.alerts.forEach((alert, index) => {
        if (alert.id === payload) {
          state.alerts.splice(index, 1)
        }
      })
    }
  }
})

export const { startAlert, removeAlert } = alertsSlice.actions
export const alertsSelector = state => state.alerts
export default alertsSlice.reducer
