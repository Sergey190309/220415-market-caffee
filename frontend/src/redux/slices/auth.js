import { createSlice } from '@reduxjs/toolkit'

import { axiosCommonToken, authAxiosClient } from '../../api/apiClient'
import { LOG_IN_INFO } from '../constants/localStorageVariables'

export const notLoggedInfo = {
  /**
   * Used for create and return to 'empty' set of state values
   */
  user_name: '',
  email: '',
  isAdmin: false,
  access_token: '',
  refresh_token: ''
}

export const logInInfo = () => {
  /**
   * Used to avoin undefined value if local ctorage does not contain
   * needed variables.
   */
  const _localStorage = localStorage.getItem(LOG_IN_INFO)
    ? { ...JSON.parse(localStorage.getItem(LOG_IN_INFO)), isLoggedIn: true }
    : { ...notLoggedInfo, isLoggedIn: false }
  // console.log('auth slice, logInInfo, _localStorage ->', _localStorage);
  return _localStorage
}

export const initialState = () => ({
  ...logInInfo(),
  loading: false,
  isSignedUp: false
})

const authSlice = createSlice({
  /**
   * I tested all reducers but signUpModalClosed and logInModalClosed.
   */
  name: 'auth',
  initialState: initialState(),
  reducers: {
    setState: (state, { payload }) => {
      // Object.keys(state).forEach(key => {
      //   delete state[key]
      // })
      Object.assign(state, payload)
    },
    signUpStart: state => {
      state.loading = true
      state.isSignedUp = false
      state.isLoggedIn = false
    },
    signUpSuccess: state => {
      localStorage.removeItem(LOG_IN_INFO)
      Object.assign(state, notLoggedInfo, {
        loading: false,
        isSignedUp: true,
        isLoggedIn: false
      })
    },
    signUpFail: state => {
      localStorage.removeItem(LOG_IN_INFO)
      Object.assign(state, notLoggedInfo, {
        isSignedUp: false,
        isLoggedIn: false,
        loading: false
      })
    },
    logInStart: state => {
      state.loading = true
      state.isSignedUp = false
      state.isLoggedIn = false
    },
    logInSuccess: (state, { payload }) => {
      // payload.isAuthenticated = true;
      // console.log('authSlice, logInSuccess, payload ->', payload);
      axiosCommonToken(payload.access_token, authAxiosClient)
      localStorage.setItem(LOG_IN_INFO, JSON.stringify(payload))
      Object.assign(state, payload, { loading: false, isLoggedIn: true })
    },
    logInFail: state => {
      localStorage.removeItem(LOG_IN_INFO)
      Object.assign(state, notLoggedInfo, {
        loading: false,
        isSignedUp: false,
        isLoggedIn: false
      })
    },
    logOut: state => {
      // console.log('authSlice, logOut')
      localStorage.removeItem(LOG_IN_INFO)
      Object.assign(state, notLoggedInfo, {
        loading: false,
        isSignedUp: false,
        isLoggedIn: false
      })
    },
    signUpModalClosed: state => {
      state.isSignedUp = false
    },
    logInModalClosed: state => {
      state.isLoggedIn = false
    }
  }
})

export const {
  setState,
  signUpStart,
  signUpSuccess,
  signUpFail,
  logInStart,
  logInSuccess,
  logInFail,
  logOut,
  signUpModalClosed,
  logInModalClosed
} = authSlice.actions

export const authSelector = state => state.auth

export default authSlice.reducer
