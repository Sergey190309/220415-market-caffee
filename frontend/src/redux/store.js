import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit'
import createSagaMiddleware from 'redux-saga'
// import reducer from './reducers';
import reducer from './slices'
import rootSaga from './saga'
// import { rootReducer } from './reducers';

const initialState = {}
const sagaMiddleware = createSagaMiddleware()

const middleware = [...getDefaultMiddleware({ thunk: false }), sagaMiddleware]
// const middleware = [sagaMiddleware, thunk];

const store = configureStore({
  reducer,
  middleware,
  initialState
})

sagaMiddleware.run(rootSaga)

export default store
