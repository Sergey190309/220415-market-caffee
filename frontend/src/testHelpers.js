import React from 'react'

import { render } from '@testing-library/react'
import { Provider } from 'react-redux'

// import store from './redux/store'

export const testRender = (jsx, { store, ...otherOpts }) => {
  return render(
    <Provider store={store}>
      {jsx}
    </Provider>,
    otherOpts
  )
}
// export function testRender (jsx, { store, ...otherOpts }) {
//   return render(<Provider store={store}>{jsx}</Provider>, otherOpts)
// }
