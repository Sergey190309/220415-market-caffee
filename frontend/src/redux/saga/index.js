import { all } from 'redux-saga/effects'
import { logInSaga, signUpSaga } from './auth'
import { alertSaga } from './alerts'
import { startInitSaga, techInSaga, lngsSaga, i18nSaga } from './tech'
import { structureSaga } from './structure'

export default function * rootSaga () {
  yield all([
    alertSaga(),
    logInSaga(),
    signUpSaga(),
    startInitSaga(),
    techInSaga(),
    lngsSaga(),
    i18nSaga(),
    structureSaga()
  ])
}
