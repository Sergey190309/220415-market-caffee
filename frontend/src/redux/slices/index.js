import { combineReducers } from '@reduxjs/toolkit'
// import { combineReducers } from 'redux';
// import alerts from '../reducers/alert';
import alerts from './alerts'
// import logIn from '../reducers/auth';
import auth from './auth'
import device from './device'
// import device from '../reducers/device';
import lng from './lng'
// import lng from '../reducers/lng';
import structure from './structure'
import tech from './tech'

export const rootReducer = combineReducers({
  alerts,
  auth,
  // logIn,
  device,
  lng,
  structure,
  tech
})

export { startAlert, removeAlert } from './alerts'
export {
  setState as authSetState,
  signUpStart,
  signUpSuccess,
  signUpFail,
  logInStart,
  logInSuccess,
  logInFail,
  logOut,
  signUpModalClosed,
  logInModalClosed,
  authSelector
} from './auth'
export { deviceSelector, setDeviceSize, openModal, closeModal, setEditable } from './device'
export { lngSwitch, lngSelector } from './lng'
export {
  setState as structureSetState,
  structureSelector,
  structureStart,
  structureSuccess,
  structureFail
} from './structure'
export {
  techSelector,
  startInitLoading,
  initLoadingSuccess,
  startTechIn,
  techInSuccess,
  techInFail,
  startLngs,
  lngsSuccess,
  lngsFail,
  startI18n,
  i18nInitiated,
  i18nSuccess,
  i18nFail
} from './tech'
export default rootReducer
