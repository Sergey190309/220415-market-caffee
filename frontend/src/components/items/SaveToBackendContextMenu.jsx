import React, { useState, useEffect } from 'react'
import { useDispatch } from 'react-redux'
import PropTypes from 'prop-types'
import {
  Menu,
  Message,
  Popup
} from 'semantic-ui-react'
import { useTranslation } from 'react-i18next'

import { ELEMENT_EDIT_SAVE, ELEMENT_EDIT_CANCEL } from '../../redux/constants/menuKeys'
import { positiveColor, warningColor } from '../../utils/colors'
import { backendTxtUpdateStart, backendTxtUpdateClear } from '../../redux/slices'
import { ContextMenuItem } from './menu_items/ContextMenuItem'

export const SaveToBackendContextMenu = t => ([

])

const SaveToBackendContextMenu = ({
  isOpened,
  context,
  setContextMenuOpened
}) => {
  const { t } = useTranslation('context')

  const [menuStructure, setMenuStructure] = useState([])

  const dispatch = useDispatch()

  useEffect(() => {
    setMenuStructure([
      {
        name: ELEMENT_EDIT_SAVE,
        icon: { name: 'save', color: positiveColor },
        content: t('2LE.saveEditedElement'),
        onClick: onClickHandler
      },
      {
        name: ELEMENT_EDIT_CANCEL,
        icon: { name: 'cancel', color: warningColor },
        content: t('2LE.cancelEditedElement'),
        onClick: onClickHandler
      }
    ])
  }, [])

  const onClickHandler = (event, { name }) => {
    event.preventDefault()
    switch (name) {
      case ELEMENT_EDIT_SAVE:
        console.log('SaveToBackendContextMenu:\n onClickHandler\n  ELEMENT_EDIT_SAVE')
        dispatch(backendTxtUpdateStart())
        break
      case ELEMENT_EDIT_CANCEL:
        console.log('SaveToBackendContextMenu:\n onClickHandler\n  ELEMENT_EDIT_CANCEL')
        dispatch(backendTxtUpdateClear())
        break
      default:
        break
    }
    setContextMenuOpened(false)
  }

  return (
    <Popup
      data-testid='Popup'
      wide
      open={isOpened}
      context={context}
      onClose={() => {
        // setOpened(false)
        setContextMenuOpened(false)
      }}
    >
      <Message
        // as='p'
        content={t('note.SaveToBackendContextMenu')}
        icon={{ name: 'exclamation', color: warningColor }}
      />
      <Menu
        vertical
        compact
      >
        {menuStructure.map((item, index) => (
          <ContextMenuItem key={index} {...item} />
        ))}
      </Menu>
    </Popup>
  )
}

SaveToBackendContextMenu.defaultProps = {
  isOpened: true,
  context: {},
  setContextMenuOpened: () => { },
  saveToBackend: () => { }

}

SaveToBackendContextMenu.propTypes = {
  isOpened: PropTypes.bool.isRequired,
  context: PropTypes.object.isRequired,
  setContextMenuOpened: PropTypes.func.isRequired,
  saveToBackend: PropTypes.func.isRequired
}

export default SaveToBackendContextMenu
