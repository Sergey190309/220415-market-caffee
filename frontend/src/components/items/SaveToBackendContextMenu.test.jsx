import React from 'react'
import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import store from '../../redux/store'
import SaveToBackendContextMenu, { saveToBackendContextMenu }
  from './SaveToBackendContextMenu'
import { makeTestStore, testRender } from '../../testHelpers'
import { backendTxtUpdateClear, backendTxtUpdateStart } from '../../redux/slices'
// import { testRender, makeTestStore } from '../../../../testHelpers'
// import store from '../../../../redux/store'
// import moduleName from '../../'

jest.mock('../../api/calls/content')

describe('saveToBackendContextMenu testing', () => {
  const testProps = {
    isOpened: true,
    context: {},
    setContextMenuOpened: jest.fn()
  }

  const renderRedux = (
    actualProps,
    testStore = makeTestStore(store)
  ) => {
    // console.log('ParagraphContextMenu.test:\n renderReduxContext',
    //   '\n  index ->', index)
    testRender(
      <SaveToBackendContextMenu {...actualProps} />
      , { store: testStore }
    )
  }
  describe('appearance', () => {
    test('rendering, elements qnt, (screenshot)', async () => {
      const acitveProps = { ...testProps }
      await waitFor(() => {
        renderRedux(acitveProps, store)
      })
      const popup = screen.getByTestId('Popup')
      expect(popup).toMatchSnapshot()
      const menu = screen.getByTestId('Menu')
      expect(menu.children)
        .toHaveLength(saveToBackendContextMenu(jest.fn()).length)
      // console.log('SaveToBackendContextMenu.test:\n rendering',
      //   '\n  popup ->', menu.children.length)
      // screen.debug(popup)
    })
    test('not opened -> no popup visible', async () => {
      const acitveProps = { ...testProps, isOpened: false }
      await waitFor(() => {
        renderRedux(acitveProps, store)
      })
      expect(screen.queryByTestId('Popup')).not.toBeInTheDocument()
      // screen.debug()
    })
  })
  describe('actions', () => {
    test('2LE.saveEditedElement', async () => {
      const acitveProps = { ...testProps }
      await waitFor(() => {
        renderRedux(acitveProps)
      })
      const save = screen.getByText('2LE.saveEditedElement')
      userEvent.click(save)
      expect(store.dispatch).toHaveBeenCalledTimes(1)
      expect(store.dispatch).toHaveBeenCalledWith(backendTxtUpdateStart())
      expect(acitveProps.setContextMenuOpened).toHaveBeenCalledTimes(1)
      expect(acitveProps.setContextMenuOpened).toHaveBeenCalledWith(false)
      // console.log('SaveToBackendContextMenu.test:\n rendering',
      //   '\n  dispatch ->', store.dispatch.mock.calls)
      // screen.debug(save)
    })
    test('2LE.cancelEditedElement', async () => {
      const acitveProps = { ...testProps }
      await waitFor(() => {
        renderRedux(acitveProps)
      })
      const cancel = screen.getByText('2LE.cancelEditedElement')
      userEvent.click(cancel)
      expect(store.dispatch).toHaveBeenCalledTimes(1)
      expect(store.dispatch).toHaveBeenCalledWith(backendTxtUpdateClear())
      expect(acitveProps.setContextMenuOpened).toHaveBeenCalledTimes(1)
      expect(acitveProps.setContextMenuOpened).toHaveBeenCalledWith(false)
      // console.log('SaveToBackendContextMenu.test:\n rendering',
      //   '\n  dispatch ->', store.dispatch.mock.calls)
      // screen.debug()
    })
  })
})
