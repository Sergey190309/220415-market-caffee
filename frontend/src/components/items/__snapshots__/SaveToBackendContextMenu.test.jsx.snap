// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`saveToBackendContextMenu testing appearance rendering, elements qnt, (screenshot) 1`] = `
<div
  class="ui top left wide popup transition visible"
  data-testid="Popup"
  style="position: initial;"
>
  <div
    class="ui icon message"
  >
    <i
      aria-hidden="true"
      class="orange exclamation icon"
    />
    <div
      class="content"
    >
      <p>
        note.SaveToBackendContextMenu
      </p>
    </div>
  </div>
  <div
    class="ui compact vertical menu"
    data-testid="Menu"
  >
    <a
      class="item sc-bdfBQB eVLmMB"
    >
      <i
        aria-hidden="true"
        class="teal save icon"
      />
      2LE.saveEditedElement
    </a>
    <a
      class="item sc-bdfBQB eVLmMB"
    >
      <i
        aria-hidden="true"
        class="orange cancel icon"
      />
      2LE.cancelEditedElement
    </a>
  </div>
</div>
`;
